File: spark_scala/src/main/scala/com/spark/datasource/scala/lesson/JSONDataSource.scala
package com.spark.datasource.scala.lesson

import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SparkSession}
import org.apache.spark.{SparkConf, SparkContext}

/**
 * @date 2021/5/12
 */
object JSONDataSource {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName("JSONDataSource").setMaster("local")
    val sc = new SparkContext(conf)
    val sqlContext = SparkSession.builder().config(conf).getOrCreate()

    // 创建学生成绩的DataSet
    val studentDS = sqlContext.read.format("json").load("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/json/students.json")
    // 创建临时表
    studentDS.createOrReplaceTempView("students")
    // 获取优秀学生的信息
    val studentScoreDS = sqlContext.sql("select name, score from students")
    studentScoreDS.printSchema()
    studentScoreDS.show()

    // 获取RDD数据
    val studentScoreRDD = studentScoreDS.rdd
    val goodStudentNames = studentScoreRDD.map(studentScore => studentScore(0)).collect()

    println(goodStudentNames)

    // 创建学生基本信息的数据
    // 创建学生基本信息DataFrame
    val studentInfoJSONs = Array(
      "{\"name\":\"Leo\", \"age\":18}",
      "{\"name\":\"Marry\", \"age\":17}",
      "{\"name\":\"Jack\", \"age\":19}")
    val studentInfoRDD = sc.parallelize(studentInfoJSONs, 3)
    val studentInfoDS = sqlContext.read.json(studentInfoRDD)

    // 查询现在分数大于80分学生的基本信息
    studentInfoDS.createOrReplaceTempView("student_infos")
    var sql = "select name,age from student_infos where name in ("
    for(i <- 0 until goodStudentNames.length) {
      sql += "'" + goodStudentNames(i) + "'"
      if(i < goodStudentNames.length - 1) {
        sql += ","
      }
    }
    sql += ")"

    // sql查询基本信息
    val goodStudentInfoRDD = sqlContext.sql(sql).rdd

    // join数据
    val joinRDD = studentScoreRDD
      .map(row => (row.getAs[String]("name"), row.getAs[Long]("score")))
      .join(goodStudentInfoRDD.map(row => (row.getAs[String]("name"), row.getAs[Long]("age"))))

    val joinRowsRDD = joinRDD.map(info => Row(info._1, info._2._1.toInt, info._2._2.toInt))

    // 类型创建
    val structType = StructType(Array(
      StructField("name", StringType, true),
      StructField("score", IntegerType, true),
      StructField("age", IntegerType, true)))

    val goodStudentsDS = sqlContext.createDataFrame(joinRowsRDD, structType)

    goodStudentsDS.show()

  }

}

File: spark_java/src/main/java/com/spark/datasource/study/JsonDataSource.java
package com.spark.datasource.study;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SQLContext;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import scala.Int;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;

/**
 * @date 2021/5/12
 */
public class JsonDataSource {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setMaster("local").setAppName("JsonDataSource");
        JavaSparkContext sc = new JavaSparkContext(conf);

        SQLContext sqlContext = new SQLContext(sc);

        Dataset<Row> studentScoreDS = sqlContext.read().json("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/json/students.json");
        studentScoreDS.createOrReplaceTempView("student_scores");

        // 学生的成绩DataSet
        Dataset<Row> goodStudentScoresDS = sqlContext.sql("select name, score from student_scores where score >= 80");

        List<String> goodStudentNames = goodStudentScoresDS.javaRDD().map(new Function<Row, String>() {
            private static final long serialVersionUID = 2285732451218399708L;
            @Override
            public String call(Row row) throws Exception {
                return row.getString(0);
            }
        }).collect();

        // 针对包含json传的javaRDD创建DataSet
        List<String> studentInfoJSONS = new ArrayList<String>();
        studentInfoJSONS.add("{\"name\":\"Leo\", \"age\":18}");
        studentInfoJSONS.add("{\"name\":\"Marry\", \"age\":17}");
        studentInfoJSONS.add("{\"name\":\"Jack\", \"age\":19}");

        JavaRDD<String> studentInfoJSONsRDD = sc.parallelize(studentInfoJSONS);
        Dataset<Row> studentInfosDS = sqlContext.read().json(studentInfoJSONsRDD);

        // 针对学生基本信息构建临时表
        studentInfosDS.createOrReplaceTempView("student_infos");

        String sql = "select name, age from student_infos where name in (";

        for (int i = 0; i < goodStudentNames.size(); i++) {
            sql += "'" + goodStudentNames.get(i) + "'";
            if (i < goodStudentNames.size() - 1) {
                sql += ",";
            }
        }
        sql += ")";
        // 学生基本信息的DataSet
        Dataset<Row> goodsStudentInfosDS = sqlContext.sql(sql);

        goodsStudentInfosDS.printSchema();
        goodsStudentInfosDS.show();

        JavaPairRDD<String, Tuple2<Integer, Integer>> goodStudentsRDD =
                goodStudentScoresDS.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
                    private static final long serialVersionUID = -5282421837742849450L;

                    @Override
                    public Tuple2<String, Integer> call(Row row) throws Exception {
                        return new Tuple2<String, Integer>(
                                row.getString(0),
                                Integer.valueOf(String.valueOf(row.getLong(1)))
                        );
                    }
                }).join(goodsStudentInfosDS.javaRDD().mapToPair(new PairFunction<Row, String, Integer>() {
                    private static final long serialVersionUID = -3482294033491512059L;

                    @Override
                    public Tuple2<String, Integer> call(Row row) throws Exception {
                        return new Tuple2<String, Integer>(
                                row.getString(0),
                                Integer.valueOf(String.valueOf(row.getLong(1)))
                        );
                    }
                }));
        // 然后将封装在RDD中的好学生的全部信息，转换为一个JavaRDD<Row>的格式
        // （将JavaRDD，转换为DataFrame）
        JavaRDD<Row> goodStudentRowsRDD = goodStudentsRDD.map(

                new Function<Tuple2<String,Tuple2<Integer,Integer>>, Row>() {

                    private static final long serialVersionUID = 1L;

                    @Override
                    public Row call(
                            Tuple2<String, Tuple2<Integer, Integer>> tuple)
                            throws Exception {
                        return RowFactory.create(tuple._1, tuple._2._1, tuple._2._2);
                    }

                });

        // 创建一份元数据，将JavaRDD<Row>转换为DataFrame
        List<StructField> structFields = new ArrayList<StructField>();
        structFields.add(DataTypes.createStructField("name", DataTypes.StringType, true));
        structFields.add(DataTypes.createStructField("score", DataTypes.IntegerType, true));
        structFields.add(DataTypes.createStructField("age", DataTypes.IntegerType, true));
        StructType structType = DataTypes.createStructType(structFields);

        Dataset<Row> goodStudentsDF = sqlContext.createDataFrame(goodStudentRowsRDD, structType);

        goodStudentsDF.printSchema();
        goodStudentsDF.show();

        // 将好学生的全部信息保存到一个json文件中去
        // （将DataFrame中的数据保存到外部的json文件中去）
        goodStudentsDF.write().format("json").save("/Users/qian/WorkSpaces/own-workspace/2021/spark_code_repository/spark_java/src/main/resources/json/good-students");

    }

}
